doc/design/concurrent_data_transfer_and_kernel_execution.md

```diff
@@ -0,0 +1,97 @@
+The staging area has one fixed-size buffer and two methods, `put` and `get`. `put` and `get` access buffer in a thread-safe way. The operation of `put` puts data into the buffer, and `get` removes a data from the buffer. When the program runs, the staging area should be warmed up. At first, `put` is called to put some data in the staging area. After that, `get` is called when the kernel begins executing, and one data is extracted from the staging area. At the same time, `put` is called and a new data is copied from CPU side to the staging area. Because the operation of data transfer is asynchronous, so the `put` method will return immediately.     
+
+The GPU provided by NVIDIA generally has two engines: a copy engine and a computing engine. To make two engines working at the same time, we need to use the [stream mechanism](https://devblogs.nvidia.com/parallelforall/gpu-pro-tip-cuda-7-streams-simplify-concurrency/) provided by CUDA. As long as data transfer and kernel execution are in different streams and the kernel does not use the data being transferred, the data transfer and kernel execution can be done at the same time on GPU.  
+
+To support the above description, we need to define a new class: `Channel`. Here, we borrow the concept of [`Channel`](https://www.golang-book.com/books/intro/10#section2) in the go language.
+
+```
+template <typename T>
+class Channel {
+ private:
+  using ChannelElement = std::vector<T>;
+
+  std::size_t capacity_;
+  std::size_t bytes_limit_;
+  std::size_t current_bytes_;
+  std::mutex mu_;
+  std::condition_variable empty_cond_var_;
+  std::condition_variable full_cond_var_;
+  std::deque<ChannelElement> channel_;
+
+ public:
+  explicit Channel(std::size_t capacity, std::size_t bytes_limit);
+
+  void Put(ChannelElement* buffer_element) { ... }
+  void Get(ChannelElement* buffer_element) { ... }
+  size_t Size() { ... }
+  void Clear()  { ... }
+};
+};
+```
+
+**Take the [recognize_digits_mlp](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_recognize_digits_mlp.py) as an example to show the model definition after using the staging area mechanism.**
+
+```
+...
+concurrent_program = Program()
+with program(concurrent_program):
+    image = fluid.layers.data_layer(...)
+    label = fluid.layers.data_layer(...)
+    places = get_places()
+    channel_list = create_channel_list(name="data_buffer")
+    with parallel.for(places) as for_loop:
+        cur_channel = fluid.layers.create_channel(channel_list, for_loop.i, channel_size=2)  
+        # if channel_list has the specific channel, 
+        # this function will return it directly.
+        fluid.layers.send_to_channel(input=[for_loop.input(image), 
+                                for_loop.input(label)], out=cur_channel)
+
+main_program = Program()
+with program(main_program):
```

Comment:
Please refer to https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/test_parallel_op.py#L79 for current usage of `parallel_do` op.